File: math/primality tests/Ultra Prime.java
problem link-https://www.hackerearth.com/practice/math/number-theory/primality-tests/practice-problems/algorithm/ultra-prime/

import java.io.*;
import java.util.*;

class TestClass {
    public static void main(String args[] ) throws Exception {
        BufferedReader bu=new BufferedReader(new InputStreamReader(System.in));
        int t=Integer.parseInt(bu.readLine());
        StringBuilder sb=new StringBuilder();
        boolean sieve[]=new boolean[1000001];
        int i,j,c=0,count[]=new int[1000001];
        
        for(i=2;i<=1000000;i++)
        if(sieve[i]==false)
        {
            for(j=2*i;j<=1000000;j+=i)
            sieve[j]=true;
        }
        
        ArrayList<Integer> pr=new ArrayList<>();
        for(i=2;i<=1000;i++)
        if(sieve[i]==false) pr.add(i);
        
        for(i=2;i<=1000000;i++)
        {
            if(sieve[i]==false)
            {
                int sum=0,no=i;
                while(no>0)
                {sum+=no%10;
                no/=10;}
                if(pr.contains(sum)) c++;
            }
            count[i]=c;
        }
        
        while(t-->0)
        {
            String s[]=bu.readLine().split(" ");
            int l=Integer.parseInt(s[0])-1,r=Integer.parseInt(s[1]);
            sb.append(count[r]-count[l]+"\n");
        }
        System.out.print(sb);
    }
}

File: math/primality tests/Ultra Prime.cpp
problem link-https://www.hackerearth.com/practice/math/number-theory/primality-tests/practice-problems/algorithm/ultra-prime/description/?layout=old

#include<bits/stdc++.h>
using namespace std;
int main()
{
	ios::sync_with_stdio(false);
	cin.tie(NULL);
	int i,j,N=1000001,pr[N+1]={0},cnt=0,c[N+1],t;
	for(i=2;i<=N;i++)
	if(pr[i]==0)
	for(j=2*i;j<=N;j+=i)
	pr[j]=1;

	c[0]=c[1]=0;
	for(i=2;i<=N;i++)
	{
		c[i]=cnt;
		if(pr[i]==0)
		{
			int sum=0,d=i;
			while(d!=0)
			{sum+=d%10; d/=10;}
			if(pr[sum]==0) cnt++;
		}
	}
	
	scanf("%d",&t);
	while(t--)
	{
		int l,r;
		scanf("%d%d",&l,&r);
		printf("%d\n",c[r+1]-c[l]);
	}
}
